# Разбор 17 задачи (уровень: Базовый)

В файле содержится последовательность целых чисел. Элементы последовательности могут принимать целые значения от –10 000 до 10 000 включительно. Определите количество пар последовательности, в которых только одно число оканчивается на 3, а сумма квадратов элементов пары не меньше квадрата максимального элемента последовательности, оканчивающегося на 3. В ответе запишите два числа: сначала количество найденных пар, затем максимальную из сумм квадратов элементов таких пар. В данной задаче под парой подразумевается два идущих подряд элемента последовательности.

Файлы к заданию: https://kompege.ru/files/8QlTUIh3y.txt

По сравнению с прошлой задачей нам нужно проверять, на что заканчиваются числа, а также находить сумму квадратов.

В этот раз я покажу новый способ открытия файла и преобразования строчных объектов в числовые.

In [23]:
file = [int(x) for x in open('17_4705.txt').readlines()]
print(file)

[5196, -2074, -1943, 3612, 4303, -9995, 2257, 9386, -7653, -235, 436, 9738, 983, -4140, -4607, 7587, -9689, -4856, 841, 1868, -2005, 716, -6879, -610, -7945, -3495, 9371, -7560, -1988, -6742, 7280, -7499, 1340, 1885, -4842, -4277, 4918, 6834, -6211, 8670, 1710, -7607, 2930, -1607, 7881, -1351, 6538, 8913, -2082, 6513, 5944, -6632, -6698, 7155, 7164, -8711, 4478, 4015, -3095, 3067, 8172, 7663, -424, -4791, -2047, 3683, -9818, 2396, -1792, -1355, -5597, 8768, 8337, -4376, -8764, 1400, -2984, -4173, 5698, -7506, 1619, -2029, 2553, 2583, -9167, 5446, -6785, -1967, -4966, -1350, -9486, -6028, 7261, 6102, 6490, -1030, 7738, -13, 9674, 4485, -7639, -6452, -5679, 2138, -7942, -7923, -3211, 99, 2127, 3834, 2211, 2927, 3369, 8498, 6113, 1128, -2512, 4977, 6979, 2253, 3726, 1472, -1010, -9760, -7631, -7813, 8096, 3401, -9079, 3411, 1125, 8311, 1198, 5181, 8925, -559, 548, -4366, -943, 4916, 4716, 2731, 7005, 1699, 5375, 3186, 6104, -5988, 5810, 3735, 2105, 9652, -3699, -3984, 3672, -8726, -7889, 

Теперь обратим внимание на основные условия поиска пар:

**"Определите количество пар последовательности, в которых только одно число оканчивается на 3, а сумма квадратов элементов пары не меньше квадрата максимального элемента последовательности, оканчивающегося на 3."**

**Найдём максимальный элемент последовательности, оканчивающийся на 3:**

In [3]:
mx = max(filter(lambda x: x % 10 == 3, file))
print(mx)

9973


Максимальный элемент последовательности, оканчивающийся на 3, найден: **9973**

Теперь возведём его в квадрат, как говорится в условии выше.

In [4]:
mx = mx ** 2
print(mx)

99460729


Теперь посмотрим, как формируются пары элементов. В данной задаче под парой подразумевается два идущих подряд элемента последовательности.

**Сформируем их:**

In [22]:
for i in range(len(file)-1):
    num1 = file[i]
    num2 = file[i+1]
    print(num1,num2)


5196 -2074
-2074 -1943
-1943 3612
3612 4303
4303 -9995
-9995 2257
2257 9386
9386 -7653
-7653 -235
-235 436
436 9738
9738 983
983 -4140
-4140 -4607
-4607 7587
7587 -9689
-9689 -4856
-4856 841
841 1868
1868 -2005
-2005 716
716 -6879
-6879 -610
-610 -7945
-7945 -3495
-3495 9371
9371 -7560
-7560 -1988
-1988 -6742
-6742 7280
7280 -7499
-7499 1340
1340 1885
1885 -4842
-4842 -4277
-4277 4918
4918 6834
6834 -6211
-6211 8670
8670 1710
1710 -7607
-7607 2930
2930 -1607
-1607 7881
7881 -1351
-1351 6538
6538 8913
8913 -2082
-2082 6513
6513 5944
5944 -6632
-6632 -6698
-6698 7155
7155 7164
7164 -8711
-8711 4478
4478 4015
4015 -3095
-3095 3067
3067 8172
8172 7663
7663 -424
-424 -4791
-4791 -2047
-2047 3683
3683 -9818
-9818 2396
2396 -1792
-1792 -1355
-1355 -5597
-5597 8768
8768 8337
8337 -4376
-4376 -8764
-8764 1400
1400 -2984
-2984 -4173
-4173 5698
5698 -7506
-7506 1619
1619 -2029
-2029 2553
2553 2583
2583 -9167
-9167 5446
5446 -6785
-6785 -1967
-1967 -4966
-4966 -1350
-1350 -9486
-9486 -6028
-6028 7

Теперь определим пары, в которых **только один элемент** заканчивается на 3.

In [19]:
for i in range(len(file)-1):  # Цикл по всем соседним парам элементов в списке
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    bool_num1 = num1 % 10 == 3  # Проверяем, оканчивается ли первое число на 3
    bool_num2 = num2 % 10 == 3  # Проверяем, оканчивается ли второе число на 3
    if bool_num1 + bool_num2 == 1:  # Проверяем, что ровно одно число оканчивается на 3
        print(num1,num2)

3612 4303
4303 -9995
9738 983
983 -4140
-4140 -4607
-4607 7587
-4842 -4277
-4277 4918
1710 -7607
-7607 2930
2930 -1607
-1607 7881
6538 8913
8913 -2082
-2082 6513
6513 5944
8172 7663
7663 -424
-4791 -2047
3683 -9818
-1355 -5597
-5597 8768
-2029 2553
-9167 5446
-6785 -1967
-1967 -4966
8498 6113
6113 1128
6979 2253
2253 3726
8109 -9827
-9827 -4242
6200 -9667
-9667 -8981
-7292 -487
-487 4596
-6431 -5177
-5177 -4975
4881 1073
1073 5275
3908 7123
7123 -7629
-2243 1793
1793 385
-9579 8363
8363 -7523
5821 7743
7743 370
-3913 5663
5663 -6536
-2304 8883
8883 2649
-8215 4573
4573 -5966
-2084 7873
7873 -5240
-5240 -8627
-9337 -2005
-145 -2897
-2897 7286
-369 7673
7673 -3464
6090 -697
4433 2001
-9492 813
813 8064
-6958 -9757
-9757 -6398
-4659 2093
2093 2398
2398 6563
1633 7911
7911 -9637
-9637 -4126
9605 -4437
-4437 2058
7939 -4627
-957 9967
-7345 -8697
-8697 -6473
5694 7593
7593 5579
3832 6723
6723 8177
6828 5233
4393 -4110
-3140 -8187
-8187 9000
9517 -1467
-1467 -3540
-7169 -4457
-4457 -3746
-579

`num % 10` - получает последнюю цифру числа (остаток от деления на 10)

`num % 10 == 3` - проверяет, равна ли последняя цифра 3

`bool_num1` и `bool_num2` - логические переменные (True/False), которые становятся True, если число оканчивается на 3

---

`bool_num1 + bool_num2 == 1` - в Python True = 1, False = 0, поэтому:

- `0 + 0 = 0` - оба числа НЕ оканчиваются на 3 ❌
- `1 + 0 = 1` - первое оканчивается на 3, второе нет ✅  
- `0 + 1 = 1` - первое не оканчивается, второе оканчивается на 3 ✅
- `1 + 1 = 2` - оба числа оканчиваются на 3 ❌

---

**Обратите внимание на 5 пару**, которая получилась: `-4140 -4607`. Как видно, ни одно число не заканчивается на 3. Но почему же они тут?

В Python есть проблемы с проверкой на окончание отрицательных чисел.

In [7]:
print(-4140 % 10)
print(-4607 % 10)

0
3


Как видно, есть проблемы со вторым числом. Почему же так происходит?

**Математическое объяснение:**
В Python оператор `%` возвращает остаток от деления, а не математический модуль. Для отрицательных чисел это работает по формуле:
`a % b = остаток`, где `0 ≤ остаток < b`

**Разберем наш пример:**
`-4607 % 10`

**Шаг 1:** Находим, сколько раз 10 помещается в -4607
`-4607 ÷ 10 = -460.7`

Берем целую часть в меньшую сторону: `-461` (это важно!)
Округление в меньшую сторону гарантирует, что остаток всегда будет неотрицательным и меньше делителя по модулю, что соответствует математическому определению операции взятия остатка в Python.

**Шаг 2:** Вычисляем по формуле:
`остаток = делимое - (делитель * целая_часть)`
`остаток = -4607 - (10 * -461)`
`остаток = -4607 - (-4610)`
`остаток = -4607 + 4610`
`остаток = 3`

**Для получения адекватного остатка при делении** добавим `abs()` для получения модуля числа.

In [20]:
for i in range(len(file)-1):  # Цикл по всем соседним парам элементов в списке
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    bool_num1 = abs(num1) % 10 == 3  # Проверяем, оканчивается ли ПОСЛЕДНЯЯ ЦИФРА ЧИСЛА на 3 (учитывая модуль)
    bool_num2 = abs(num2) % 10 == 3  # Проверяем, оканчивается ли ПОСЛЕДНЯЯ ЦИФРА ЧИСЛА на 3 (учитывая модуль)
    if bool_num1 + bool_num2 == 1:  # Проверяем, что ровно одно число оканчивается на 3
        print(num1, num2)

-2074 -1943
-1943 3612
3612 4303
4303 -9995
9386 -7653
-7653 -235
9738 983
983 -4140
6538 8913
8913 -2082
-2082 6513
6513 5944
8172 7663
7663 -424
-2047 3683
3683 -9818
-2984 -4173
-4173 5698
-2029 2553
2583 -9167
7738 -13
-13 9674
-7942 -7923
-7923 -3211
8498 6113
6113 1128
6979 2253
2253 3726
-7631 -7813
-7813 8096
-4366 -943
-943 4916
-8261 -4233
-4233 2891
1266 -8123
-8123 5605
3536 -13
-13 -6431
8592 -3503
-3503 -628
4881 1073
1073 5275
3908 7123
7123 -7629
-3094 -2243
1793 385
2300 -9013
-9013 -8496
-9579 8363
-7523 5821
5821 7743
7743 370
370 -3913
5663 -6536
-3048 -483
-483 8804
-2304 8883
8883 2649
-8215 4573
4573 -5966
-2084 7873
7873 -5240
-3004 -6123
-6123 -6456
9089 -7633
-7633 -5625
5035 -2383
-2343 -2715
-369 7673
7673 -3464
-3178 -5803
-5803 -2072
-697 4433
4433 2001
-9492 813
813 8064
-4659 2093
2093 2398
2398 6563
1633 7911
7489 -1973
-1973 6360
-8697 -6473
-6473 5694
5694 7593
7593 5579
3832 6723
6723 8177
8177 -63
-63 7698
2579 -1003
-1003 -7171
6828 5233
4393 -4110

`num % 10` - возвращает остаток от деления на 10, который всегда равен последней цифре числа

`abs(num) % 10` - возвращает последнюю цифру абсолютного значения числа

---

Теперь каждая наша пара имеет ровно одно число, оканчивающееся на 3. Продолжим выполнять условия, следующие -> **"а сумма квадратов элементов пары не меньше квадрата максимального элемента последовательности"**. Получим сумму квадратов элементов пары и найдём их количество.

In [16]:
c = 0  # Счетчик пар, удовлетворяющих условиям
for i in range(len(file)-1):  # Цикл по всем соседним парам элементов
    num1 = file[i]           # Текущий элемент последовательности
    num2 = file[i+1]         # Следующий элемент последовательности
    bool_num1 = abs(num1) % 10 == 3  # Проверяет, оканчивается ли последняя цифра num1 на 3
    bool_num2 = abs(num2) % 10 == 3  # Проверяет, оканчивается ли последняя цифра num2 на 3
    sm2 = num1**2 + num2**2  # Вычисляет сумму квадратов двух чисел
    if bool_num1 + bool_num2 == 1 and sm2 >= mx:  # Проверяет два условия:
        # 1. Ровно одно число оканчивается на 3
        # 2. Сумма квадратов не меньше значения переменной mx
        c += 1  # Увеличиваем счетчик подходящих пар
print(c)  # Выводим количество найденных пар

180


Перейдём ко второму вопросу. **"В ответе запишите два числа: сначала количество найденных пар, затем максимальную из сумм квадратов элементов таких пар."**

In [18]:
c = 0  # Счетчик пар, удовлетворяющих условиям
sum_mx = 0  # Переменная для хранения максимальной суммы квадратов среди подходящих пар
for i in range(len(file)-1):  # Цикл по всем соседним парам элементов списка
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    bool_num1 = abs(num1) % 10 == 3  # Проверяем, оканчивается ли num1 на 3 (с учетом модуля)
    bool_num2 = abs(num2) % 10 == 3  # Проверяем, оканчивается ли num2 на 3 (с учетом модуля)
    sm2 = num1**2 + num2**2  # Вычисляем сумму квадратов двух чисел
    if bool_num1 + bool_num2 == 1 and sm2 >= mx:  # Проверяем условия:
        # 1. Ровно одно число оканчивается на 3
        # 2. Сумма квадратов не меньше значения переменной mx
        c += 1               # Увеличиваем счетчик подходящих пар
        sum_mx = max(sum_mx, sm2)  # Обновляем максимальную сумму квадратов
print(c, sum_mx)  # Выводим количество пар и максимальную сумму квадратов

180 190360573


`sum_mx = max(sum_mx, sm2)` - находит максимальную сумму квадратов среди всех пар, удовлетворяющих условиям

**Итоговый ответ:** 180 190360573

**Полный код:**

In [21]:
# Читаем файл и преобразуем все строки в список целых чисел
file = [int(x) for x in open('17_4705.txt').readlines()]

# Находим максимальное число в файле, которое оканчивается на 3
mx = max(filter(lambda x: x % 10 == 3, file))

# Возводим это максимальное число в квадрат
mx = mx ** 2

c = 0  # Счетчик пар, удовлетворяющих условиям
sum_mx = 0  # Переменная для хранения максимальной суммы квадратов среди подходящих пар

for i in range(len(file)-1):  # Цикл по всем соседним парам элементов списка
    num1 = file[i]           # Получаем текущий элемент последовательности
    num2 = file[i+1]         # Получаем следующий элемент последовательности
    bool_num1 = abs(num1) % 10 == 3  # Проверяем, оканчивается ли num1 на 3 (с учетом модуля)
    bool_num2 = abs(num2) % 10 == 3  # Проверяем, оканчивается ли num2 на 3 (с учетом модуля)
    sm2 = num1**2 + num2**2  # Вычисляем сумму квадратов двух чисел
    
    # Проверяем условия:
    # 1. Ровно одно число оканчивается на 3
    # 2. Сумма квадратов не меньше квадрата максимального числа, оканчивающегося на 3
    if bool_num1 + bool_num2 == 1 and sm2 >= mx:
        c += 1               # Увеличиваем счетчик подходящих пар
        sum_mx = max(sum_mx, sm2)  # Обновляем максимальную сумму квадратов

# Выводим количество пар и максимальную сумму квадратов
print(c, sum_mx)

180 190360573
